Add wasSchemeOptimisticallyUpgraded in ResourceRequest

sysrqb opened the pull request:
#### 88e183bf643e6301eab37233672d0b77032b7e0a
<pre>
Add wasSchemeOptimisticallyUpgraded in ResourceRequest
<a href="https://bugs.webkit.org/show_bug.cgi?id=277009">https://bugs.webkit.org/show_bug.cgi?id=277009</a>
<a href="https://rdar.apple.com/132411852">rdar://132411852</a>

Reviewed by Charlie Wolfe.

This flag tracks whether the scheme was previously conditionally upgraded. This
will help us handle upgrades correctly, and it will help us make decisions
about how we handle request failures.

* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestResource):
* Source/WebCore/platform/network/ResourceRequestBase.cpp:
(WebCore::ResourceRequestBase::upgradeInsecureRequest):
(WebCore::ResourceRequestBase::upgradeInsecureRequestIfNeeded):
(WebCore::ResourceRequestBase::setWasSchemeOptimisticallyUpgraded):
* Source/WebCore/platform/network/ResourceRequestBase.h:
(WebCore::ResourceRequestBase::RequestData::RequestData):
(WebCore::ResourceRequestBase::wasSchemeOptimisticallyUpgraded const):
* Source/WebCore/platform/network/cf/ResourceRequest.h:
* Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
(WebCore::ResourceRequest::ResourceRequest):
(WebCore::ResourceRequest::getResourceRequestPlatformData const):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/281351@main">https://commits.webkit.org/281351@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d66ea3ace35cd4c3276fdc7bbf70eb71908b1bb8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59570 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38915 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12094 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63485 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10093 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61699 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46568 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10246 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48336 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7069 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61600 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36327 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51564 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29172 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33031 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8816 "Found 1 new test failure: imported/w3c/web-platform-tests/mathml/relations/css-styling/padding-border-margin/padding-border-margin-003.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9017 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54979 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9094 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65217 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3498 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8989 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55677 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3509 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51557 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55799 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2904 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8903 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34729 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35812 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36898 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35557 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebCore/platform/network/ResourceRequestBase.cpp:
```diff
@@ -224,7 +224,12 @@ void ResourceRequestBase::upgradeInsecureRequestIfNeeded(URL& url, ShouldUpgrade
+    if (url != m_requestData.m_url) {
+        m_requestData.m_url = url;
+        setwasSchemeOptimisticallyUpgraded(true);
+    }
```
nit - this could just be `setWasSchemeOptimisticallyUpgraded(url != m_requestData.m_url)`.

Or you could make `upgradeInsecureRequestIfNeeded()` return whether or not the scheme was upgraded to avoid copying the URL.

charliewolfe commented on Source/WebCore/platform/network/ResourceRequestBase.h:
```diff
@@ -280,6 +282,9 @@ class ResourceRequestBase {
     void setIsPrivateTokenUsageByThirdPartyAllowed(bool);
 
+    bool wasSchemeOptimisticallyUpgraded() const { return m_requestData.m_wasSchemeOptimisticallyUpgraded; }
+    void setwasSchemeOptimisticallyUpgraded(bool);
```
```suggestion
    void setWasSchemeOptimisticallyUpgraded(bool);
```

sysrqb replied:
Good catch, that was a bad replace-all substitution.

charliewolfe reviewed (commented):
Why wasSchemeOptimisticallyUpgraded instead of just wasSchemeUpgraded?

charliewolfe commented on Source/WebCore/platform/network/ResourceRequestBase.h:
```diff
@@ -111,6 +112,7 @@ class ResourceRequestBase {
         bool m_useAdvancedPrivacyProtections : 1 { false };
         bool m_didFilterLinkDecoration : 1 { false };
         bool m_isPrivateTokenUsageByThirdPartyAllowed : 1 { false };
+        bool m_wasSchemeOptimisticallyUpgraded: 1 { false };
```
```suggestion
        bool m_wasSchemeOptimisticallyUpgraded : 1 { false };
```

charliewolfe commented on Source/WebCore/platform/network/ResourceRequestBase.cpp:
```diff
@@ -201,30 +201,33 @@ void ResourceRequestBase::upgradeInsecureRequest(URL& url)
 void ResourceRequestBase::upgradeInsecureRequestIfNeeded(ShouldUpgradeLocalhostAndIPAddress shouldUpgradeLocalhostAndIPAddress, const std::optional<uint16_t>& upgradePort)
 {
-    upgradeInsecureRequestIfNeeded(m_requestData.m_url, shouldUpgradeLocalhostAndIPAddress, upgradePort);
+    auto wasUpgraded = upgradeInsecureRequestIfNeeded(m_requestData.m_url, shouldUpgradeLocalhostAndIPAddress, upgradePort);
```
nit - `bool` is just as short as `auto`

Commit:
Add wasSchemeOptimisticallyUpgraded in ResourceRequest
https://bugs.webkit.org/show_bug.cgi?id=277009
rdar://132411852

Reviewed by Charlie Wolfe.

This flag tracks whether the scheme was previously conditionally upgraded. This
will help us handle upgrades correctly, and it will help us make decisions
about how we handle request failures.

* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestResource):
* Source/WebCore/platform/network/ResourceRequestBase.cpp:
(WebCore::ResourceRequestBase::upgradeInsecureRequest):
(WebCore::ResourceRequestBase::upgradeInsecureRequestIfNeeded):
(WebCore::ResourceRequestBase::setWasSchemeOptimisticallyUpgraded):
* Source/WebCore/platform/network/ResourceRequestBase.h:
(WebCore::ResourceRequestBase::RequestData::RequestData):
(WebCore::ResourceRequestBase::wasSchemeOptimisticallyUpgraded const):
* Source/WebCore/platform/network/cf/ResourceRequest.h:
* Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
(WebCore::ResourceRequest::ResourceRequest):
(WebCore::ResourceRequest::getResourceRequestPlatformData const):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/281351@main
